src/seraphis_wallet/key_container.cpp

```diff
@@ -0,0 +1,223 @@
+// paired header
+#include "key_container.h"
+
+// local headers
+#include "crypto/chacha.h"
+#include "misc_log_ex.h"
+#include "ringct/rctOps.h"
+#include "seraphis_core/jamtis_core_utils.h"
+#include "seraphis_core/sp_core_enote_utils.h"
+#include "jamtis_keys.h"
+#include "seraphis_wallet/encrypted_file.h"
+
+// standard headers
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "seraphis_wallet"
+
+namespace seraphis_wallet
+{
+//-------------------------------------------------------------------------------------------------------------------
+// INTERNAL
+static jamtis_keys ser_keys_to_jamtis(ser_jamtis_keys const &keys)
+{
+    return jamtis_keys{
+        .k_m = keys.k_m,
+        .k_vb = keys.k_vb,
+        .xk_ua = keys.xk_ua,
+        .xk_fr = keys.xk_fr,
+        .s_ga = keys.s_ga,
+        .s_ct = keys.s_ct,
+        .K_1_base = keys.K_1_base,
+        .xK_ua = keys.xK_ua,
+        .xK_fr = keys.xK_fr,
+    };
+}
+//-------------------------------------------------------------------------------------------------------------------
+// INTERNAL
+static KeyContainer ser_container_to_usable(ser_KeyContainer const &key_container, const crypto::chacha_key &key)
+{
+    return KeyContainer(ser_keys_to_jamtis(key_container.keys), key_container.encrypted, key_container.encryption_iv);
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool jamtis_keys_valid(jamtis_keys const &keys)
+{
+    jamtis_keys test_keys{keys};
+
+    sp::jamtis::make_jamtis_unlockamounts_key(test_keys.k_vb, test_keys.xk_ua);
+    sp::jamtis::make_jamtis_findreceived_key(test_keys.k_vb, test_keys.xk_fr);
+    sp::jamtis::make_jamtis_generateaddress_secret(test_keys.k_vb, test_keys.s_ga);
+    sp::jamtis::make_jamtis_ciphertag_secret(test_keys.s_ga, test_keys.s_ct);
+    sp::make_seraphis_spendkey(test_keys.k_vb, test_keys.k_m, test_keys.K_1_base);
+    sp::jamtis::make_jamtis_unlockamounts_pubkey(test_keys.xk_ua, test_keys.xK_ua);
+    sp::jamtis::make_jamtis_findreceived_pubkey(test_keys.xk_fr, test_keys.xK_ua, test_keys.xK_fr);
+
+    return test_keys == keys;
+}
+//-------------------------------------------------------------------------------------------------------------------
+KeyContainer::KeyContainer(jamtis_keys &&keys, const crypto::chacha_key &key)
+    : m_keys{keys}, m_encrypted{false}, m_encryption_iv{}
+{
+    encrypt(key);
+}
+//-------------------------------------------------------------------------------------------------------------------
+KeyContainer::KeyContainer(jamtis_keys &&keys, bool encrypted, const crypto::chacha_iv encryption_iv)
+    : m_keys{keys}, m_encrypted{encrypted}, m_encryption_iv{encryption_iv}
+{
+}
+//-------------------------------------------------------------------------------------------------------------------
+boost::optional<KeyContainer> KeyContainer::load_from(const std::string &path, const crypto::chacha_key &key)
```

Comment:
Right now I don't think that we will ever have to support this use case in the final code, i.e. reading and writing a key container *alone* in and out of files. Maybe it's a good idea to move such capabilities into the unit tests.